entity-service-client/src/main/java/org/hypertrace/entity/query/service/client/EntityCacheKey.java

```diff
@@ -0,0 +1,39 @@
+package org.hypertrace.entity.query.service.client;
+
+import java.util.Map;
+import java.util.Objects;
+
+public class EntityCacheKey<T> {
+  private final T dataKey;
+  private final String tenantId;
+  private final Map<String, String> headers;
+
+  public EntityCacheKey(T key, String tenantId, Map<String, String> headers) {
```

Comment:
RequestContext is defined in gateway-service.

Reply:
I meant this one https://github.com/hypertrace/java-grpc-utils/blob/main/grpc-context-utils/src/main/java/org/hypertrace/core/grpcutils/context/RequestContext.java - that's generally what we're starting with right?

Reply:
Yes, but we unpack it in GatewayServiceImpl as soon as we start handling the request. Also the grpc client takes in either tenantId or headers https://github.com/hypertrace/java-grpc-utils/blob/b16a4025066defb2cf4a5f8822c3a4f428e9f7cd/grpc-client-utils/src/main/java/org/hypertrace/core/grpcutils/client/GrpcClientRequestContextUtil.java#L40 . It creates a new grpc RequestContext object from these headers or the tenantId, different from the one that came into the server. To change this, will involve loads of changes.

Reply:
So I don't really understand why we chose to do it that way. We don't have to change it, but we can easily go around it I think. For example for the Rx stuff:
https://github.com/hypertrace/java-grpc-utils/blob/b16a4025066defb2cf4a5f8822c3a4f428e9f7cd/grpc-client-rx-utils/src/main/java/org/hypertrace/core/grpcutils/client/rx/DefaultGrpcRxExecutionContext.java

I did this as a wrapper rather than putting it directly on request context so as not to introduce an Rx dependency across the platform, but other than that - we're just creating a new GRPC context, setting our own context as a key'd value in it, then using the grpc context's built in methods to make the function call. There's no need to build up a new context explicitly - we could just add `call`/`run`/`stream` methods to `RequestContext`.

That said, if it makes sense to do it the new way we should deprecate and transition the old stuff, but that can be tech debt.

Reply:
Ok. Is it a good idea to use the same requestContext that came into server to make client requests? Maybe we would have to create a new one before doing the invocation? I think for graphql it's ok since they don't use the clients defined by the services. They access the API directly.

Reply:
>  Is it a good idea to use the same requestContext that came into server to make client requests?

It should be fine (and less error prone) - the propagation filtering happens on the server call interceptor, so its already good to propagate before we even read it in a server implementation. Is there some other issue you were thinking of?

> I think for graphql it's ok since they don't use the clients defined by the services. They access the API directly.

True - although it's doing it that way because propagating context in this way makes using the generated clients just as easy as using the hand-coded clients. Context could always be propagated, dehydrated and then use the old clients in the same way as they've been used.

Does that make sense? This stuff is a bit confusing so I hope I'm not talking in circles here.